Formula/folly.rb

```diff
@@ -90,3 +95,25 @@ def install
     system "./test"
   end
 end
+
+__END__
+diff --git a/folly/detail/base64_detail/Base64Api.cpp b/folly/detail/base64_detail/Base64Api.cpp
+index b02148a64..ad074525b 100644
+--- a/folly/detail/base64_detail/Base64Api.cpp
++++ b/folly/detail/base64_detail/Base64Api.cpp
+@@ -19,13 +19,13 @@
+ #include <folly/detail/base64_detail/Base64Api.h>
+ #include <folly/detail/base64_detail/Base64SWAR.h>
+ 
+-#if FOLLY_X64
++#if FOLLY_X64 && FOLLY_SSE_PREREQ(4, 2)
+ #include <folly/detail/base64_detail/Base64_SSE4_2.h>
+ #endif
+ 
+ namespace folly::detail::base64_detail {
+ 
+-#if FOLLY_X64
++#if FOLLY_X64 && FOLLY_SSE_PREREQ(4, 2)
+ Base64RuntimeImpl base64EncodeSelectImplementation() {
+   if (folly::CpuId().sse42()) {
```

Comment:
I wonder if `folly` has CPU runtime detection if we build with SSE4.2/AVX support.

Reply:
I do see similar runtime detection code in other files like 

https://github.com/facebook/folly/blob/v2022.07.25.00/folly/hash/Checksum.cpp#L67-L75
```cpp
bool crc32c_hw_supported() {
  static folly::CpuId id;
  return id.sse42();
}

bool crc32_hw_supported() {
  static folly::CpuId id;
  return id.sse42();
}
```

https://github.com/facebook/folly/blob/v2022.07.25.00/folly/experimental/crypto/LtHash.cpp#L62-L76
```cpp
// static
template <>
bool MathOperation<MathEngine::SSE2>::isAvailable() {
  static const bool kIsAvailable =
      CpuId().sse2() && MathOperation<MathEngine::SSE2>::isImplemented();
  return kIsAvailable;
}

// static
template <>
bool MathOperation<MathEngine::AVX2>::isAvailable() {
  static const bool kIsAvailable =
      CpuId().avx2() && MathOperation<MathEngine::AVX2>::isImplemented();
  return kIsAvailable;
}
```

So, the initial macros like `FOLLY_SSE_PREREQ(4, 2)`, `__SSE2__`, `__AVX2__` may be for host compilation support while SSE/AVX may be detected at runtime.

Reply:
Did folly use your patch? They may not be willing to look at patches posted in comments since they technically need a CLA to use contributed code.

Reply:
They added a more extensive fix https://github.com/facebook/folly/commit/437b29dd062a40c741b2147827025a2f413b76c7, which is available in next release, but CMake logic is currently broken as upstream reversed x86_64 vs non-x86_64 logic https://github.com/facebook/folly/commit/9d6acb35763d329d3fd6f4223ebf134f03e65d15.

We could update to `2022.07.25.00`, patch the CMake logic, and potentially add `ENV.runtime_cpu_detection`.

Not sure what is output if we try to compile with both `-march=core2` and `-msse4.2` on Linux (I think the `-mavx` code is skipped since we don't add `libsodium` dependency).

Reply:
`-march=core2` is only an alias for the lower level flags (e.g. `-mmmx`, `-msse`, etc), so the compiler should handle it fine. But adding `-msse4.2` is almost equivalent to setting `-march=nehalem`, I think.